### PR TITLE
Bunch of cards 10

### PR DESCRIPTION
**Pull request by VJubert**

### CODE REVIEW

File: server/game/cards/02_SHD/units/RoseTicoDedicatedToTheCause.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    public override setupCardAbilities() {
+        this.addOnAttackAbility({
+            title: 'Defeat a Shield token on a friendly unit. If you do, give 2 Experience tokens to that unit.',
+            optional: true,
+            targetResolver: {
+                cardTypeFilter: WildcardCardType.Upgrade,
+                cardCondition: (card, context) => card.isShield() && card.parentCard.controller === context.source.controller,
+                immediateEffect: AbilityHelper.immediateEffects.simultaneous([
+                    // TODO: this is a hack to store the parent card of the upgrade before it's defeated.
+                    // once last known state is implemented, we need to read the upgrade's parent card from that (same for on-attack)
```
Comment (AMMayberry1) [inline server/game/cards/02_SHD/units/RoseTicoDedicatedToTheCause.ts:22]: nit: can remove the "same for on attack" part of the comment

File: test/server/cards/02_SHD/events/SparkOfHope.spec.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+                context.player2.clickCard(context.wampa);
+
+                context.player1.clickCard(context.sparkOfHope);
+                expect(context.player1).toBeAbleToSelectExactly([context.pykeSentinel, context.imperialInterceptor]);
+                context.player1.clickCard(context.pykeSentinel);
+
+                expect(context.player2).toBeActivePlayer();
+                expect(context.pykeSentinel).toBeInZone('resource');
+                expect(context.imperialInterceptor).toBeInZone('discard');
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/events/SparkOfHope.spec.ts:33]: Can we break this test into two phases with one unit defeated each phase to confirm that a unit defeated in the previous phase doesn't work

File: test/server/cards/02_SHD/leaders/HunterOutcastSergeant.spec.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                        leader: 'hunter#outcast-sergeant',
+                    },
+                    player2: {
+                        resources: ['echo#restored']
+                    }
+                });
+
+
+                const { context } = contextRef;
+                const set2Falcon = context.player1.findCardByName('millennium-falcon#landos-pride', 'resource');
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/leaders/HunterOutcastSergeant.spec.ts:20]: nit: could we name this `resourceFalcon` to make the test easier to read

File: test/server/cards/02_SHD/leaders/HunterOutcastSergeant.spec.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                context.player1.clickCard(context.hunter);
+
+                // only cards which share a name with friendly unique unit
+                expect(context.player1).toBeAbleToSelectExactly([set2Falcon, resourceEcho]);
+                expect(context.player1).toHaveChooseNoTargetButton();
+
+                context.player1.clickCard(set2Falcon);
+
+                expect(context.player2).toBeActivePlayer();
+                expect(context.player1.resources.length).toBe(4);
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/leaders/HunterOutcastSergeant.spec.ts:32]: Can we declare some deck cards explicitly so we can confirm that the specific card is moved? Same for deployed test

File: test/server/cards/02_SHD/leaders/HunterOutcastSergeant.spec.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                contextRef.setupTest({
+                    phase: 'action',
+                    player1: {
+                        groundArena: ['battlefield-marine', 'echo#restored'],
+                        spaceArena: ['millennium-falcon#piece-of-junk'],
+                        resources: ['millennium-falcon#landos-pride', 'echo#restored', 'battlefield-marine', 'devotion'],
+                        leader: 'hunter#outcast-sergeant',
+                    },
+                    player2: {
+                        resources: ['echo#restored']
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/leaders/HunterOutcastSergeant.spec.ts:14]: Can we add a unique card that is inthe opponent's arena as well to confirm it's not included in the targeting? Could just give player 2 like a deployed Leia leader and put the Leia unit in resources for player 1

File: test/server/cards/02_SHD/units/GideonsLightCruiserDarkTroopersStation.spec.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                // unit should be played as free
+                expect(context.player2).toBeActivePlayer();
+                expect(context.scoutBikePursuer).toBeInZone('groundArena');
+                expect(context.player1.exhaustedResourceCount).toBe(8);
+
+                // reset
+                context.player1.moveCard(context.gideonsLightCruiser, 'hand');
+                context.player2.passAction();
+
+                // play gideon's light cruiser again, phase 3 trooper should be played automatically
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/units/GideonsLightCruiserDarkTroopersStation.spec.ts:100]: For this one could we actually do it in the reverse order and confirm that if there is no target in discard and a target in hand, it lets you choose to do nothing in line with the hidden information rule. You probably would need to use `.toHavePassSingleTargetPrompt()`, can see an example of that in Alliance Dispatcher tests

File: test/server/cards/02_SHD/units/RoseTicoDedicatedToTheCause.spec.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                    player2: {
+                        groundArena: [{ card: 'atst', upgrades: ['shield'] }]
+                    }
+                });
+            });
+
+            it('should defeat a shield on a friendly unit and give 2 experience tokens', function () {
+                const { context } = contextRef;
+                const myShields = context.player1.findCardsByName('shield');
+                const roseTicoShield = myShields.find((s) => s.parentCard === context.roseTico);
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/units/RoseTicoDedicatedToTheCause.spec.ts:24]: Could do this as a one-liner `const roseTicoShield = context.roseTico.upgrades.filter((card) => card.isShield())[0];`

File: test/server/cards/02_SHD/units/RoseTicoDedicatedToTheCause.spec.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+                context.player1.clickCard(context.roseTico);
+                context.player1.clickCard(context.p2Base);
+
+                // able to select all shield on friendly unit
+                expect(context.player1).toBeAbleToSelectExactly(myShields);
```
Comment (AMMayberry1) [inline test/server/cards/02_SHD/units/RoseTicoDedicatedToTheCause.spec.ts:30]: Would be good to add a friendly shield on an opponent unit using Lom Pyke and confirm that the targeting still works right (can see a very similar case in the Lom Pyke tests)

File: test/server/cards/02_SHD/units/GideonsLightCruiserDarkTroopersStation.spec.ts (hunk 6 above)
Comment (VJubert) [reply test/server/cards/02_SHD/units/GideonsLightCruiserDarkTroopersStation.spec.ts]: @AMMayberry1 did you reopen this comment ? I reverse order on the second test. So I have a test with pass option in hand and no pass option on discard

File: test/server/cards/02_SHD/events/SparkOfHope.spec.ts [hunk 9]
```diff
@@ -32,6 +32,25 @@ describe('Spark of Hope', function () {
                 expect(context.pykeSentinel).toBeInZone('resource');
                 expect(context.imperialInterceptor).toBeInZone('discard');
             });
+
+            it('can resource a unit defeated this phase (do not play unit defeated previous phases)', function () {
```
[the remark this answers is not in the document]
Comment (VJubert) [reply test/server/cards/02_SHD/events/SparkOfHope.spec.ts:36]: tests are slightly different, the second could be "play the last defeated unit"

**Pull request merged**